### ISSUE TITLE
Learning a sampler for arbitrary Potts/Ising models

### ISSUE THREAD

**Issue by alex-hh** [alex-hh/project-ideas#5]
Sampling from these models is hard...but they have a specific graph structure...
so maybe we should generate a bunch of models, and train a conditional GAN or an MMD GAN
or something to sample from them.

The conditioning can be done by feeding the parameters through a GNN.

(It'd be much easier to just find the closest autoregressive model, which implies
a graph-to-graph model plus an ordering predictor.)

**Comment by alex-hh** [comment]
## Simulations:
### GWG (N.B. they basically copy Ingraham, just using different language)
[Paper](https://arxiv.org/pdf/2102.04509.pdf)
We generate Ising models with different sparse graph structures; a 2D cyclic lattice and a random Erdos-Renyi graph.
We generate training data with a long-run Gibbs chain and
train models using Persistent Contrastive Divergence (Tieleman, 2008) with an `1 penalty to encourage sparsity. We
evaluate our models using the Root-Mean-Squared-Error (RMSE) between the inferred connectivity matrix Jˆ and the
true matrix J.

### Ingraham
[Paper](http://proceedings.mlr.press/v70/ingraham17a/ingraham17a.pdf)
[Supplement](http://proceedings.mlr.press/v70/ingraham17a/ingraham17a-supp.pdf)
#### Ising
We generated two synthetic systems. The first system was
ferromagnetic (all J ≥ 0) with 64 spins, where neighboring spins xi
, xj have a nonzero interaction of Jij = 0.2
if adjacent on a 4 × 4 × 4 periodic lattice. This coupling
strength equates to being slightly above the critical temperature, meaning the system will be highly correlated despite
the underlying interactions being only nearest-neighbor.
The second system was a diluted Sherrington-Kirkpatrick
spin glass (Sherrington & Kirkpatrick, 1975; Aurell & Ekeberg, 2012) with 100 spins. The couplings in this model
were defined by Erdos-Renyi random graphs ( ˝ Erdos & ˝
Renyi ´ , 1960) with non-zero edge weights distributed as
Jij ∼ N 
0,
1
Np
where N p is the average degree. We
generated 5 random systems where the average degree was
N p = 100(0.02) = 2. Across all of the systems, we used
Swendsen-Wang sampling (Swendsen & Wang, 1987) to
sample synthetic data and checked that the sampling was
sufficient to eliminate autocorrelation in the data.

For the ER graph, we create a model with 200 nodes. The ER edge probability is chosen so each node has an average
of 4 neighbors. The strength of each edge is IID sampled from N(0,1/4)

#### Potts
We constructed a synthetic Potts spin glass with sparse interactions chosen to reflect an underlying 3D structure. After forming a contact topology from a random packed polymer, we generated synthetic group-Student-t distributed
sitewise bias vectors hi (each 20 × 1) and Gaussian distributed coupling matrices Jij (each 20 × 20) to mirror
the strong site-bias and weak-coupling regime of proteins.
Since this system is highly frustrated, we thinned 2 × 106
sweeps of Gibbs sampling to 2000 sequences that exhibited
no autocorrelation.
Given 400 of the 2000 synthetic sequences1
, we inferred
L2 and group L1-regularized MAP estimates under a pseudolikelihood approximation with 5-fold cross validation
to choose hyperparameters from 6 values in the range
{0.3, 1.0, 3.0, 10.0, 30.0, 100.0}. We also ran PVI-10 with
40 persistent Markov chains and 5000 iterations of stochastic gradient descent with Adam2
(Kingma & Ba, 2014). We
note that the current standards of the field are based on L2


**Comment by alex-hh** [comment]
v0 of this: learn a cluster sampler by maximising acceptance probability.
